fix(resolve): update the ambiguity glob binding recursively

bvanjoi opened the pull request:
Fixes #112713

The error was not reported for #112713 because the `ambiguity` in the binding of `root::sub::C` was lost.

The `resolve_imports` process is shown below:

|Iter| `importer` | `module` | `resolve_glob_import` |
| -  | - | - | - |
| 0  | `use sub::*`| `root::sub`         | Since no bindings had been added, nothing was defined|
| 1  | `use mod1::*` | `root::sub::mod1` | 1. `(root::sub, C) -> root::sub::mod1::C` by `try_define`; <br/>2. `(block_module, C) -> NameBinding { kind: Import { binding: root::sub::mod1::C }, .. }` by `try_define` recursively for `module.glob_importers`  |
| 2  | `use mod2::*` | `root::sub::mod2` | `(root::sub, C) -> root::sub::mod1::C` with  ambiguity  |

However, the ambiguity binding generated by iteration 2 points to a different value than the one generated by iteration 1. Therefore, the binding to ﻿`root::sub::mod::C` in iteration 1 does not have ambiguity, and no error is thrown in the end.

In this PR, I had made update the ambiguity glob binding recursively.

Perhaps changing `﻿&'a NameBinding<'a>` to ﻿`&'a mut NameBinding<'a>` in `﻿NameResolution` would be a better choice.

r? @petrochenkov

bvanjoi commented:
Additionally, I am unsure whether fixing errors such as "it should throw an error, but it's not" constitutes a breaking change or not.

bors commented:
:umbrella: The latest upstream changes (presumably #112774) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented on compiler/rustc_resolve/src/imports.rs:
```diff
@@ -321,6 +321,11 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                         } else if !old_binding.vis.is_at_least(binding.vis, this.tcx) {
```
```suggestion
                        } else if binding.is_ambiguity() || !old_binding.vis.is_at_least(binding.vis, this.tcx) {
```
The `matches!(..., NameBindingKind::Import { .. })` conditions are redundant, we are already checking it above.

bvanjoi replied:
Oh yes, I just realized that I had already run ﻿`binding.is_glob_import`.

petrochenkov commented:
>Additionally, I am unsure whether fixing errors such as "it should throw an error, but it's not" constitutes a breaking change or not.

It's a breaking change, but it is also a bugfix.
Could you rebase this PR so I could run crater to estimate the amount of breakage?
@rustbot author

petrochenkov commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit 1576b8ed41c10aa2263e55f774217b07c8c17966 with merge b1375b2484512d46eeace16657e5b4a5450c610b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"1576b8ed41c10aa2263e55f774217b07c8c17966","merge_sha":"b1375b2484512d46eeace16657e5b4a5450c610b"} -->

rust-timer commented:
Queued b1375b2484512d46eeace16657e5b4a5450c610b with parent 18a6d911caba59605eb03db1452848a85d2e5879, future [comparison URL](https://perf.rust-lang.org/compare.html?start=18a6d911caba59605eb03db1452848a85d2e5879&end=b1375b2484512d46eeace16657e5b4a5450c610b).
<!-- rust-timer: temporary -->

petrochenkov commented:
@bors rollup=maybe

bvanjoi committed:
fix(resolve): update the ambiguity glob binding recursively

bvanjoi commented on tests/ui/resolve/issue-56593-2.rs:
```diff
@@ -0,0 +1,26 @@
+
+    #[test]
+    fn test_thing() {
+        let thing = Thing::Foo;
```
This is a breaking change too: `root::Thing` currently refers to `root::thing::Thing`, but it will refer to `root::thing::Thing` after this PR.

petrochenkov replied:
The comment has a typo, but I assume you mean `root::thing::Thing` -> `root::Thing`.
It's also clearly a bugfix.

Crater will finish soon and show what actually breaks after these changes.

bvanjoi commented:
Update:

I have rebased and updated this PR. I would like to declare that this PR may have a much impact, as it:
- Fixes #56593 and introduces a breaking change, see this [comment](https://github.com/rust-lang/rust/pull/112743?notification_referrer_id=NT_kwDOAcyhV7M2Nzk2ODExNDUzOjMwMTg3ODYz#discussion_r1236924481).
- Fixes #105235.
- Fixes #112713 and also introduces a breaking change

bvanjoi commented:
I think that removing ﻿`_ if old_binding() => return t` is the correct fix, but it may cause significant compatibility issues. Should I explore alternative fixes?

bvanjoi commented on tests/ui/resolve/issue-112713.rs:
```diff
@@ -0,0 +1,16 @@
+
+pub fn foo() -> u32 {
+    use sub::*;
+    C //~ERROR `C` is ambiguous
```
This is a breaking change: This is a breaking change. It will throw an ambiguous error after this PR, but it currently compiles without error.

petrochenkov commented:
Crater report generation is stuck because the number of regressions is huge.
The very popular crate `openssl` in particular is broken with this PR.
Could you check why is it broken?
@rustbot author

petrochenkov commented:
`openssl` errors:
```
[INFO] fetching crate openssl 0.10.54...
[INFO] checking openssl-0.10.54 against try#b1375b2484512d46eeace16657e5b4a5450c610b for pr-112743
[INFO] extracting crate openssl 0.10.54 into /workspace/builds/worker-4-tc2/source
[INFO] validating manifest of crates.io crate openssl 0.10.54 on toolchain b1375b2484512d46eeace16657e5b4a5450c610b
[INFO] running `Command { std: CARGO_HOME="/workspace/cargo-home" RUSTUP_HOME="/workspace/rustup-home" "/workspace/cargo-home/bin/cargo" "+b1375b2484512d46eeace16657e5b4a5450c610b" "metadata" "--manifest-path" "Cargo.toml" "--no-deps", kill_on_drop: false }`
[INFO] started tweaking crates.io crate openssl 0.10.54
[INFO] finished tweaking crates.io crate openssl 0.10.54
[INFO] tweaked toml for crates.io crate openssl 0.10.54 written to /workspace/builds/worker-4-tc2/source/Cargo.toml
[INFO] crate crates.io crate openssl 0.10.54 already has a lockfile, it will not be regenerated
[INFO] running `Command { std: CARGO_HOME="/workspace/cargo-home" RUSTUP_HOME="/workspace/rustup-home" "/workspace/cargo-home/bin/cargo" "+b1375b2484512d46eeace16657e5b4a5450c610b" "fetch" "--manifest-path" "Cargo.toml", kill_on_drop: false }`
[INFO] running `Command { std: "docker" "create" "-v" "/var/lib/crater-agent-workspace/builds/worker-4-tc2/target:/opt/rustwide/target:rw,Z" "-v" "/var/lib/crater-agent-workspace/builds/worker-4-tc2/source:/opt/rustwide/workdir:ro,Z" "-v" "/var/lib/crater-agent-workspace/cargo-home:/opt/rustwide/cargo-home:ro,Z" "-v" "/var/lib/crater-agent-workspace/rustup-home:/opt/rustwide/rustup-home:ro,Z" "-e" "SOURCE_DIR=/opt/rustwide/workdir" "-e" "CARGO_TARGET_DIR=/opt/rustwide/target" "-e" "CARGO_HOME=/opt/rustwide/cargo-home" "-e" "RUSTUP_HOME=/opt/rustwide/rustup-home" "-w" "/opt/rustwide/workdir" "-m" "1610612736" "--user" "0:0" "--network" "none" "ghcr.io/rust-lang/crates-build-env/linux@sha256:19e73879b68f455e41552ea35293514db52ea292d993f571636dbffedb77a1ee" "/opt/rustwide/cargo-home/bin/cargo" "+b1375b2484512d46eeace16657e5b4a5450c610b" "metadata" "--no-deps" "--format-version=1", kill_on_drop: false }`
[INFO] [stdout] f22bca74cbb476d0038c4cff311be5ca469a556887b7cb82fe4a475162d09a7f
[INFO] running `Command { std: "docker" "start" "-a" "f22bca74cbb476d0038c4cff311be5ca469a556887b7cb82fe4a475162d09a7f", kill_on_drop: false }`
[INFO] running `Command { std: "docker" "inspect" "f22bca74cbb476d0038c4cff311be5ca469a556887b7cb82fe4a475162d09a7f", kill_on_drop: false }`
[INFO] running `Command { std: "docker" "rm" "-f" "f22bca74cbb476d0038c4cff311be5ca469a556887b7cb82fe4a475162d09a7f", kill_on_drop: false }`
[INFO] [stdout] f22bca74cbb476d0038c4cff311be5ca469a556887b7cb82fe4a475162d09a7f
[INFO] running `Command { std: "docker" "create" "-v" "/var/lib/crater-agent-workspace/builds/worker-4-tc2/target:/opt/rustwide/target:rw,Z" "-v" "/var/lib/crater-agent-workspace/builds/worker-4-tc2/source:/opt/rustwide/workdir:ro,Z" "-v" "/var/lib/crater-agent-workspace/cargo-home:/opt/rustwide/cargo-home:ro,Z" "-v" "/var/lib/crater-agent-workspace/rustup-home:/opt/rustwide/rustup-home:ro,Z" "-e" "SOURCE_DIR=/opt/rustwide/workdir" "-e" "CARGO_TARGET_DIR=/opt/rustwide/target" "-e" "CARGO_INCREMENTAL=0" "-e" "RUST_BACKTRACE=full" "-e" "RUSTFLAGS=--cap-lints=forbid" "-e" "RUSTDOCFLAGS=--cap-lints=forbid" "-e" "CARGO_HOME=/opt/rustwide/cargo-home" "-e" "RUSTUP_HOME=/opt/rustwide/rustup-home" "-w" "/opt/rustwide/workdir" "-m" "1610612736" "--user" "0:0" "--network" "none" "ghcr.io/rust-lang/crates-build-env/linux@sha256:19e73879b68f455e41552ea35293514db52ea292d993f571636dbffedb77a1ee" "/opt/rustwide/cargo-home/bin/cargo" "+b1375b2484512d46eeace16657e5b4a5450c610b" "check" "--frozen" "--all" "--all-targets" "--message-format=json", kill_on_drop: false }`
[INFO] [stdout] 30571e612a9c73d6c89678478ee9952e2e14cf3669f07e3f00e017a5d0eb440d
[INFO] running `Command { std: "docker" "start" "-a" "30571e612a9c73d6c89678478ee9952e2e14cf3669f07e3f00e017a5d0eb440d", kill_on_drop: false }`
[INFO] [stderr]    Compiling proc-macro2 v1.0.54
[INFO] [stderr]    Compiling cc v1.0.78
[INFO] [stderr]    Compiling unicode-ident v1.0.6
[INFO] [stderr]    Compiling quote v1.0.26
[INFO] [stderr]    Compiling pkg-config v0.3.26
[INFO] [stderr]    Compiling libc v0.2.139
[INFO] [stderr]    Compiling openssl v0.10.54 (/opt/rustwide/workdir)
[INFO] [stderr]     Checking foreign-types-shared v0.1.1
[INFO] [stderr]     Checking once_cell v1.14.0
[INFO] [stderr]     Checking hex v0.3.2
[INFO] [stderr]     Checking foreign-types v0.3.2
[INFO] [stderr]    Compiling openssl-sys v0.9.88
[INFO] [stderr]    Compiling syn v2.0.11
[INFO] [stderr]    Compiling openssl-macros v0.1.1
[INFO] [stdout] error[E0425]: cannot find function, tuple struct or tuple variant `EVP_PKEY_id` in crate `ffi`
[INFO] [stdout]    --> src/pkey.rs:195:36
[INFO] [stdout]     |
[INFO] [stdout] 195 |         unsafe { Id::from_raw(ffi::EVP_PKEY_id(self.as_ptr())) }
[INFO] [stdout]     |                                    ^^^^^^^^^^^ help: a constant with a similar name exists: `EVP_PKEY_DH`
[INFO] [stdout]     |
[INFO] [stdout]    ::: /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.88/src/./evp.rs:11:1
[INFO] [stdout]     |
[INFO] [stdout] 11  | pub const EVP_PKEY_DH: c_int = NID_dhKeyAgreement;
[INFO] [stdout]     | ---------------------------- similarly named constant `EVP_PKEY_DH` defined here
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error[E0425]: cannot find function, tuple struct or tuple variant `EVP_PKEY_bits` in crate `ffi`
[INFO] [stdout]    --> src/pkey.rs:231:23
[INFO] [stdout]     |
[INFO] [stdout] 231 |           unsafe { ffi::EVP_PKEY_bits(self.as_ptr()) as u32 }
[INFO] [stdout]     |                         ^^^^^^^^^^^^^ help: a function with a similar name exists: `EVP_PKEY_sign`
[INFO] [stdout]     |
[INFO] [stdout]    ::: /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.88/src/./handwritten/evp.rs:538:5
[INFO] [stdout]     |
[INFO] [stdout] 538 | /     pub fn EVP_PKEY_sign(
[INFO] [stdout] 539 | |         ctx: *mut EVP_PKEY_CTX,
[INFO] [stdout] 540 | |         sig: *mut c_uchar,
[INFO] [stdout] 541 | |         siglen: *mut size_t,
[INFO] [stdout] 542 | |         tbs: *const c_uchar,
[INFO] [stdout] 543 | |         tbslen: size_t,
[INFO] [stdout] 544 | |     ) -> c_int;
[INFO] [stdout]     | |______________- similarly named function `EVP_PKEY_sign` defined here
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error[E0425]: cannot find function, tuple struct or tuple variant `EVP_PKEY_security_bits` in crate `ffi`
[INFO] [stdout]    --> src/pkey.rs:240:23
[INFO] [stdout]     |
[INFO] [stdout] 240 |         unsafe { ffi::EVP_PKEY_security_bits(self.as_ptr()) as u32 }
[INFO] [stdout]     |                       ^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `EVP_PKEY_get_security_bits`
[INFO] [stdout]     |
[INFO] [stdout]    ::: /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.88/src/./handwritten/evp.rs:414:13
[INFO] [stdout]     |
[INFO] [stdout] 414 |             pub fn EVP_PKEY_get_security_bits(key: *const EVP_PKEY) -> c_int;
[INFO] [stdout]     |             ---------------------------------------------------------------- similarly named function `EVP_PKEY_get_security_bits` defined here
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error[E0425]: cannot find function, tuple struct or tuple variant `EVP_PKEY_id` in crate `ffi`
[INFO] [stdout]    --> src/pkey.rs:195:36
[INFO] [stdout]     |
[INFO] [stdout] 195 |         unsafe { Id::from_raw(ffi::EVP_PKEY_id(self.as_ptr())) }
[INFO] [stdout]     |                                    ^^^^^^^^^^^ help: a constant with a similar name exists: `EVP_PKEY_DH`
[INFO] [stdout]     |
[INFO] [stdout]    ::: /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.88/src/./evp.rs:11:1
[INFO] [stdout]     |
[INFO] [stdout] 11  | pub const EVP_PKEY_DH: c_int = NID_dhKeyAgreement;
[INFO] [stdout]     | ---------------------------- similarly named constant `EVP_PKEY_DH` defined here
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error[E0425]: cannot find function, tuple struct or tuple variant `EVP_PKEY_bits` in crate `ffi`
[INFO] [stdout]    --> src/pkey.rs:231:23
[INFO] [stdout]     |
[INFO] [stdout] 231 |           unsafe { ffi::EVP_PKEY_bits(self.as_ptr()) as u32 }
[INFO] [stdout]     |                         ^^^^^^^^^^^^^ help: a function with a similar name exists: `EVP_PKEY_sign`
[INFO] [stdout]     |
[INFO] [stdout]    ::: /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.88/src/./handwritten/evp.rs:538:5
[INFO] [stdout]     |
[INFO] [stdout] 538 | /     pub fn EVP_PKEY_sign(
[INFO] [stdout] 539 | |         ctx: *mut EVP_PKEY_CTX,
[INFO] [stdout] 540 | |         sig: *mut c_uchar,
[INFO] [stdout] 541 | |         siglen: *mut size_t,
[INFO] [stdout] 542 | |         tbs: *const c_uchar,
[INFO] [stdout] 543 | |         tbslen: size_t,
[INFO] [stdout] 544 | |     ) -> c_int;
[INFO] [stdout]     | |______________- similarly named function `EVP_PKEY_sign` defined here
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error[E0425]: cannot find function, tuple struct or tuple variant `EVP_PKEY_security_bits` in crate `ffi`
[INFO] [stdout]    --> src/pkey.rs:240:23
[INFO] [stdout]     |
[INFO] [stdout] 240 |         unsafe { ffi::EVP_PKEY_security_bits(self.as_ptr()) as u32 }
[INFO] [stdout]     |                       ^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `EVP_PKEY_get_security_bits`
[INFO] [stdout]     |
[INFO] [stdout]    ::: /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/openssl-sys-0.9.88/src/./handwritten/evp.rs:414:13
[INFO] [stdout]     |
[INFO] [stdout] 414 |             pub fn EVP_PKEY_get_security_bits(key: *const EVP_PKEY) -> c_int;
[INFO] [stdout]     |             ---------------------------------------------------------------- similarly named function `EVP_PKEY_get_security_bits` defined here
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error: aborting due to 3 previous errors
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] For more information about this error, try `rustc --explain E0425`.
[INFO] [stdout] 
[INFO] [stderr] error: could not compile `openssl` (lib) due to 4 previous errors
[INFO] [stderr] warning: build failed, waiting for other jobs to finish...
[INFO] [stdout] error: aborting due to 3 previous errors
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] For more information about this error, try `rustc --explain E0425`.
[INFO] [stdout] 
[INFO] [stderr] error: could not compile `openssl` (lib test) due to 4 previous errors
[INFO] running `Command { std: "docker" "inspect" "30571e612a9c73d6c89678478ee9952e2e14cf3669f07e3f00e017a5d0eb440d", kill_on_drop: false }`
[INFO] running `Command { std: "docker" "rm" "-f" "30571e612a9c73d6c89678478ee9952e2e14cf3669f07e3f00e017a5d0eb440d", kill_on_drop: false }`
[INFO] [stdout] 30571e612a9c73d6c89678478ee9952e2e14cf3669f07e3f00e017a5d0eb440d
```

petrochenkov commented:
All regressions: https://crater-reports.s3.amazonaws.com/pr-112743/logs-archives/regressed.tar.gz

bvanjoi commented:
reduce: `rustc +stage1 code.rs --cfg ossl300`

```rs
macro_rules! cfg_if {
  ($(
      if #[cfg($($meta:meta),*)] { $($it:item)* }
  ) else * else {
      $($it2:item)*
  }) => {
      cfg_if! {
          @__items
          () ;
          $( ( ($($meta),*) ($($it)*) ), )*
          ( () ($($it2)*) ),
      }
  };

  (
      if #[cfg($($i_met:meta),*)] { $($i_it:item)* }
      $(
          else if #[cfg($($e_met:meta),*)] { $($e_it:item)* }
      )*
  ) => {
      cfg_if! {
          @__items
          () ;
          ( ($($i_met),*) ($($i_it)*) ),
          $( ( ($($e_met),*) ($($e_it)*) ), )*
          ( () () ),
      }
  };

  (@__items ($($not:meta,)*) ; ) => {};
  (@__items ($($not:meta,)*) ; ( ($($m:meta),*) ($($it:item)*) ), $($rest:tt)*) => {
      cfg_if! { @__apply cfg(all($($m,)* not(any($($not),*)))), $($it)* }

      cfg_if! { @__items ($($not,)* $($m,)*) ; $($rest)* }
  };

  (@__apply $m:meta, $($it:item)*) => {
      $(#[$m] $it)*
  };
}

mod openssl {
    pub use self::evp::*;
    pub use self::handwritten::*;

    mod evp {
        cfg_if! {
            if #[cfg(ossl300)] {
                pub unsafe fn EVP_PKEY_id() {}
            }
        }
    }

    mod handwritten {
        pub use self::evp::*;

        mod evp {
            cfg_if! {
                if #[cfg(ossl300)]  {
                    pub unsafe fn EVP_PKEY_id() {}
                } else {
                    extern "C" {
                        pub fn EVP_PKEY_id();
                    }
                }
            }
        }
    }
}

pub use openssl::*;

pub fn id() {
    unsafe {
        EVP_PKEY_id();
    }
}


fn main() {}
```

bvanjoi commented:
One strange thing is that when I expand it, `rustc code.rs -cfg ossl300` and `rustc +stage1 -cfg ossl300` both throw ambiguous error.

```rs
mod openssl {
    pub use self::evp::*;
    pub use self::handwritten::*;

    mod evp {
        #[cfg(all(ossl300, not(any())))]
        pub unsafe fn EVP_PKEY_id() {}
    }

    mod handwritten {
        pub use self::evp::*;

        mod evp {
            #[cfg(all(ossl300, not(any())))]
            pub unsafe fn EVP_PKEY_id() {}

            #[cfg(all(not(any(ossl300))))]
            extern "C" {
                pub fn EVP_PKEY_id();
            }
        }
    }
}
pub use openssl::*;

fn main() {
    unsafe {
        EVP_PKEY_id();
    }
}
```

petrochenkov commented:
Yes, also a bugfix.
Could you try fixing `openssl` and asking them to release a new version?
It should remove majority of those 8909 regressions and then we'll think what to do next (probably introduce deprecation warnings).

bvanjoi commented:
I have written a script to analyze this regression log and found that it contains 58 types of issues, some of which are unrelated to name resolution:
[regression.zip](https://github.com/rust-lang/rust/files/11859067/regression.zip)

petrochenkov commented:
The right fix here is probably adding a new boolean flag somewhere around `struct AmbiguityError` that would discern between ambiguities reported previously and new ambiguities added by this PR.
Then `fn report_ambiguity_error` would report a hard error for the former, and a deprecation lint for the latter.

In any case we need to wait until the new release of openssl is published, because it's responsible for the majority of regressions, and because it's not just an ambiguity error, it's a "cannot find" error, likely due to a cross-crate use of this ambiguity.

petrochenkov commented:
Superseded by https://github.com/rust-lang/rust/pull/113099.
